src/r1cs/circuits/jolt_duplicated.circom

```diff
@@ -0,0 +1,316 @@
+    signal z__4 <== z__5 + is_add_instr * z_add;
+    signal z__3 <== z__4 + is_jump_instr * z_jump;
+    signal z__2 <== z__3 + is_sub_instr * z_sub;
+    signal z__1 <== z__2 + is_mul_instr * z_mul;
+    signal z <== z__1;
+
+    // verify chunks_x
+    signal combined_x_chunks <== combine_chunks(N_CHUNKS(), L_CHUNK())(chunks_x);
+    assert((combined_x_chunks - x) * is_concat == 0);
+
+    // verify chunks_y 
+    signal combined_y_chunks <== combine_chunks(N_CHUNKS(), L_CHUNK())(chunks_y);
+    assert((combined_y_chunks-y) * is_concat == 0);
+
+    // verify chunks of z -- doesn't matter if concat, though
+    signal combined_z_chunks <== combine_chunks(N_CHUNKS(), L_CHUNK())(chunks_z);
+    assert((combined_z_chunks-z) * (1-is_concat) == 0);
+
+    // verify that chunks_i == chunks_x || chunks_y -- only for concat
+    for (var i=0; i<N_CHUNKS(); i++) {
+      assert(chunks_lookup[i] - (chunks_x[i] + chunks_y[i] * 2**L_CHUNK()) * is_concat == 0);
+    } 
+    // the most significant chunk has a shorter length!
+    var idx_ms_chunk = N_CHUNKS()-1;
+    assert(chunks_lookup[idx_ms_chunk] - (chunks_x[idx_ms_chunk] + chunks_y[idx_ms_chunk] * 2**(L_MS_CHUNK())) * is_concat == 0);
+
+    // Get final lookup index (query)
+    signal lookup_index <== op_code * (2**128) + z;
+
+    is_assert_false_instr * (1-lookup_output) === 0;
+    is_assert_true_instr * lookup_output === 0;
+
+    // lui doesn't need a lookup 
+   is_lui * (lookup_output - immediate) === 0;
+
+   /*
+        Store into destination register rd 
+   */
+    signal to_store_in_rd <== if_update_rd * lookup_output;
+    signal to_store_in_rd_j <== if_else()([is_jump_instr, lookup_output-4, to_store_in_rd]);
+
+    /*
+        Store into output state
+    */
+    output_state[STEP_NUM_IDX()] <== input_state[STEP_NUM_IDX()]+1;
+
+    // set next PC 
+    signal next_pc_j <== if_else()([is_jump_instr, lookup_output-4, input_state[PC_IDX()] + 4]);
+    signal next_pc_j_b <== if_else()([is_branch_instr * lookup_output, input_state[PC_IDX()] + sign_imm * immediate, next_pc_j]);
+    output_state[PC_IDX()] <== next_pc_j_b;
+    
+}
+
+template JoltLoop(N) {
+    var TOTAL_INPUT_LENGTH = N * WIT_PER_STEP();
+    signal input in[TOTAL_INPUT_LENGTH];
+
+
+    signal output out[2]; 
+
+    /* Variables used in the for loop
+    */
+    component jolt_steps[N];
+
+    for (var i=0; i<N; i++) {
+        jolt_steps[i] = JoltStep();
+
+        if (i==0) {
+            jolt_steps[i].input_state <== [0, 0];
+        } else {
+            jolt_steps[i].input_state <== jolt_steps[i-1].output_state;
+        }
+
+        jolt_steps[i].witnesses <== subarray(i * WIT_PER_STEP(), WIT_PER_STEP(), TOTAL_INPUT_LENGTH)(in); 
+
+    }
+
+    out <== jolt_steps[N-1].output_state;
+}
+
+template JoltMain(N) {
+    signal input in[N * WIT_PER_STEP()];
+
+    component jolt_loop; 
+    jolt_loop = JoltLoop(N); 
+    jolt_loop.in <== in;
+
+ }
+
+component main {public [in]} 
+    = JoltMain(1);
```

Comment:
Eventually I think we'll need to make this file a template which gets written by `build.rs` such that we can set the constants at runtime. Fine for now.